### PR TITLE
perf: improve diff performance by 35% in worst-case

### PR DESCRIPTION
**Pull request by haydenflinner**
.

### CODE REVIEW

**Review by joshka — CHANGES_REQUESTED** [review]
What's the absolute measurement on this change, and what percentage of the total rendering time does this take up? Was this something that you identified that was slow and it actually causes a measurable difference?

35% is a lot, but if we're talking nano seconds, there's probably better gains to be had elsewhere than struct initialization.

I'd suggest putting a benchmark somewhere higher in the stack to get a feel for how this particular part of the code path relates to the whole. An example of this is `bench_paragraph_new()` which takes pretty much no time at all in the benchmarks, but is useful to see in comparison to the much larger dominating performance factors found in scrolling and rendering scrolled values.

Overall I don't mind merging this change, and it's probably a good to start to increase the amount of benchmarks in the code, but we should lean towards mostly making performance fixes in parts of the code where there's real performance issues.

File: benches/paragraph.rs [hunk 1]
```diff
@@ -161,5 +161,48 @@ fn random_lines(count: u16) -> String {
...
-criterion_group!(benches, paragraph);
+fn criterion_benchmark(c: &mut Criterion) {
+    paragraph(c);
+    buf_diff(c);
+}
+
+/// Benchmark for rendering a paragraph with a given number of lines. The design of this benchmark
+/// allows comparison of the performance of rendering a paragraph with different numbers of lines.
+/// as well as comparing with the various settings on the scroll and wrap features.
+pub fn buf_diff(c: &mut Criterion) {
```
[the remark this answers is not in the document]
Comment (haydenflinner) [reply benches/paragraph.rs:172]: I agree, but syntactically I couldn't figure out how to have two different .rs files while also being able to do `cargo bench buf_filter` (or any variation thereof where I filter down to just the buffer tests).

File: benches/paragraph.rs [hunk 2]
```diff
@@ -161,5 +161,48 @@ fn random_lines(count: u16) -> String {
...
+                for j in 0..elem_count {
+                    let c = buf.get_mut(i, j);
+                    c.set_char(char);
+                }
+            }
+            buf
+        };
+
+        let buf1 = get_buf('a');
+        let buf2 = get_buf('b');
```
[the remark this answers is not in the document]
Comment (haydenflinner) [reply benches/paragraph.rs:192]: That's much more brief, thanks! Something interesting I learned here is that we're using a u16 for the buffer size for some reason, so 255x255 is actually as big as we can handle, 256x256 is too big.

**Comment by haydenflinner** [comment]
>What's the absolute measurement on this change, and what percentage of the total rendering time does this take up? Was this something that you identified that was slow and it actually causes a measurable difference?

image

In my example I see diff taking 16% of the each frame loop, which I thought was high given the size of the screen and how complicated all of the text layout code is. It is really slow performance for the data sizes involved due to the allocations and per-cell logic everywhere. It was probably easier to implement this way, and I can see why one might think it was optimal to minimize the work that flows out of diff, but the worst-case for diff ought to be (full screen size * 100 copies / ram speed) == 650500 bytes / 10GB/s == 0.0006505 / 10 == 650 microseconds, while ideally we'd do less than 100 copies of the complete buffer during processing and hit some cache before RAM. Instead I get 1000 micros even with this micro-optimization. To improve will require a change to the return type of the function, and probably some changes to the interface with the backends. But I think 100x  faster is easily doable, I just won't be focusing any attention here for the near future.

>I'd suggest putting a benchmark somewhere higher in the stack to get a feel for how this particular part of the code path relates to the whole. An example of this is bench_paragraph_new() which takes pretty much no time at all in the benchmarks, but is useful to see in comparison to the much larger dominating performance factors found in scrolling and rendering scrolled values.
>Overall I don't mind merging this change, and it's probably a good to start to increase the amount of benchmarks in the code, but we should lean towards mostly making performance fixes in parts of the code where there's real performance issues.

Kind of agreed, I'm actually leaning towards just deleting the benchmark since it provides a false target for anyone who comes after. The first target for performance should indeed be the text-reflowing, and towards that end the next time I sit down to look at performance it will probably be at trying to help get eyesonjune's patch to work with lifetimes so that I can swap in a simpler/faster wrapping algorithm, as I'm more alarmed by the speed of that. I'm not exactly targetting 60fps, but I was surprised to see just how much time was spent in ratatui given the amount of eyes that have looked at it and how much time I thought my code was spending. I just really like the green color criterion prints when you 'win' 😊  and I usually find the faster code simpler as well (it's generally just doing less, so less complex).

**Comment by haydenflinner** [comment]
I unfortunately couldn't convince criterion to let me use its 'save-baseline' feature either, not sure what's up with [that](https://www.google.com/search?channel=fs&client=ubuntu-sn&q=criterion+save+baseline). So here is a run with the vec capacity reverted. The 255 case used to show a 35% gain, seems to be significant noise on my laptop.
```rust
     Running benches/paragraph.rs (target/release/deps/paragraph-fd41067c9435a50c)
Gnuplot not found, using plotters backend
buf_diff/buf_diff/1     time:   [32.754 ns 32.984 ns 33.281 ns]
                        change: [+7.3910% +9.2937% +11.042%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  5 (5.00%) high severe
buf_diff/buf_diff/10    time:   [1.9282 µs 1.9593 µs 2.0022 µs]
                        change: [+30.016% +31.914% +34.161%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe
Benchmarking buf_diff/buf_diff/255: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.5s, enable flat sampling, or reduce sample count to 60.
buf_diff/buf_diff/255   time:   [1.0814 ms 1.0936 ms 1.1077 ms]
                        change: [+4.6655% +7.4879% +10.521%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe

```

**Comment by haydenflinner** [comment]
I lied, it was bugging me :smile: 

```rust
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches/paragraph.rs (target/release/deps/paragraph-fd41067c9435a50c)
Gnuplot not found, using plotters backend
buf_diff/buf_diff/1     time:   [24.641 ns 25.726 ns 26.944 ns]
                        change: [-23.152% -20.159% -17.314%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) high mild
  7 (7.00%) high severe
buf_diff/buf_diff/10    time:   [773.58 ns 779.07 ns 786.38 ns]
                        change: [-47.704% -46.295% -44.708%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe
buf_diff/buf_diff/255   time:   [662.33 µs 667.19 µs 672.33 µs]
                        change: [-36.599% -34.792% -32.715%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
```

**Comment by joshka** [comment]
Ugh - apologies for that - I pushed a rebase / fix for the issues I mentioned, but git pushed `ratatui/main` instead of `haydenflinner/main` to your fork's main branch. Doing that automatically closed the PR (and makes me unable to reopen it for you as I lose permissions on the branch as a result).

I've pushed the updated commits to https://github.com/joshka/ratatui/tree/haydenflinner/main and the results are 17%/40%/54% (on my machine - YMMV):

```
❯ git sw HEAD~ --detach  
HEAD is now at 353d68a perf(buffer): add buffer benchmark

❯ cargo bench --bench buffer -- --save-baseline main
   Compiling ratatui v0.21.0 (/Users/joshka/local/ratatui)
    Finished bench [optimized] target(s) in 1.42s
     Running benches/buffer.rs (target/release/deps/buffer-78eed67bf97b80cb)
Buffer::set_string      time:   [156.34 ns 157.12 ns 157.94 ns]
                        change: [-1.3725% -0.4467% +0.4978%] (p = 0.36 > 0.05)
                        No change in performance detected.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

Buffer::diff/1          time:   [28.814 ns 28.942 ns 29.070 ns]
                        change: [-0.5691% +0.4600% +1.4751%] (p = 0.39 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
Buffer::diff/10         time:   [1.0574 µs 1.0635 µs 1.0696 µs]
                        change: [-4.3858% -3.4079% -2.4484%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
Buffer::diff/255        time:   [511.21 µs 513.45 µs 515.70 µs]
                        change: [-2.4073% -1.6399% -0.9093%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild

❯ git sw -             
warning: refname '@{-1}' is ambiguous.
Previous HEAD position was 353d68a perf(buffer): add buffer benchmark
Switched to branch 'haydenflinner/main'

❯ cargo bench --bench buffer -- --baseline main                                    
   Compiling ratatui v0.21.0 (/Users/joshka/local/ratatui)
    Finished bench [optimized] target(s) in 0.96s
     Running benches/buffer.rs (target/release/deps/buffer-78eed67bf97b80cb)
Buffer::set_string      time:   [159.16 ns 159.86 ns 160.60 ns]
                        change: [+1.5295% +2.4471% +3.4694%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

Buffer::diff/1          time:   [23.891 ns 23.992 ns 24.100 ns]
                        change: [-18.290% -17.391% -16.577%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  3 (3.00%) high severe
Buffer::diff/10         time:   [484.00 ns 486.15 ns 488.26 ns]
                        change: [-54.699% -54.322% -53.939%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild
Buffer::diff/255        time:   [305.84 µs 306.98 µs 308.17 µs]
                        change: [-40.712% -40.364% -40.008%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe
```

I think you can probably reopen this PR by pushing some changes to your main branch again, otherwise I'll open a new PR with those changes on your behalf. Here's the diff: https://github.com/tui-rs-revival/ratatui/compare/main...joshka:ratatui:haydenflinner/main

**Comment by joshka** [comment]
reopen

**Comment by haydenflinner** [comment]
Hey @joshka , I tried clicking `Create pull request` from your diff link there, it says I must be a collaborator 😆 Please feel free to submit a PR when you get around to it, I am no good at the Github/local/fork/PR workflow

**Pull request closed without merging**